bazel/ray_deps_setup.bzl

```diff
@@ -1,80 +1,80 @@
 load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository", "new_git_repository")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
 
-def github_repository(*, name=None, remote=None, commit=None, tag=None,
-                      branch=None, build_file=None, build_file_content=None,
-                      sha256=None, archive_suffix=".zip", shallow_since=None,
-                      strip_prefix=True, url=None, path=None, **kwargs):
+def urlsplit(url):
+    """ Splits a URL like "https://example.com/a/b?c=d&e#f" into a tuple:
+        ("https", ["example", "com"], ["a", "b"], ["c=d", "e"], "f")
+    A trailing slash will result in a correspondingly empty final path component.
     """
-    Conveniently chooses between archive, git, etc. GitHub repositories.
-    Prefer archives, as they're smaller and faster due to the lack of history.
+    split_on_anchor = url.split("#", 1)
+    split_on_query = split_on_anchor[0].split("?", 1)
+    split_on_scheme = split_on_query[0].split("://", 1)
+    if len(split_on_scheme) <= 1:  # Scheme is optional
+        split_on_scheme = [None] + split_on_scheme[:1]
+    split_on_path = split_on_scheme[1].split("/")
+    return (split_on_scheme[0],
+            split_on_path[0].split("."),
+            split_on_path[1:],
+            split_on_query[1].split("&") if len(split_on_query) > 1 else None,
+            split_on_anchor[1] if len(split_on_anchor) > 1 else None)
+
+def auto_http_archive(*, name=None, url=None, urls=True,
+                      build_file=None, build_file_content=None,
+                      strip_prefix=True, **kwargs):
+    """ Intelligently choose mirrors based on the given URL for the download.
```

Comment:
one sentence about the heuristic we are using here? (when to use bazel mirror and when to use github directly?)

Reply:
Currently the heuristic is we prioritize `mirror.bazel.build` URLs below GitHub URLs, but above everything else.

Reply:
(The line you're looking for is `prefer_url_over_mirrors = is_github`.)